File: apps/api/app/api/v1/routers/employees.py
"""The AI company, exposed.

Read endpoints describe the workforce (registry, capabilities, tools, health);
the write endpoint hands a goal to the Orchestrator. The frontend never hardcodes
the roster -- it renders whatever the registry reports.
"""

import uuid

from fastapi import APIRouter, HTTPException, Query, status
from pydantic import BaseModel

from app.core.dependencies import CurrentUser, DB
from app.employees import capabilities as caps
from app.employees import memory as memory_layer
from app.employees import orchestrator, registry, toolbelt

router = APIRouter()


class DelegateRequest(BaseModel):
    goal: str
    project_id: uuid.UUID
    persona: str = "creator"
    tier: str | None = None
    capabilities: list[str] | None = None
    parallelism: int = orchestrator.DEFAULT_PARALLELISM


class MemoryWrite(BaseModel):
    project_id: uuid.UUID | None = None
    employee_id: str
    content: str
    scope: str = "project"
    kind: str = "note"
    key: str | None = None


# --- the workforce ------------------------------------------------------------


@router.get("")
async def list_employees(
    current_user: CurrentUser,
    department: str | None = Query(default=None),
    capability: str | None = Query(default=None),
    include_disabled: bool = Query(default=False),
) -> dict:
    """The registry. This is the single source of truth for who works here."""
    if capability:
        employees = registry.find_by_capability(capability)
    else:
        employees = registry.all_employees(include_disabled=include_disabled)
    if department:
        employees = [e for e in employees if e.department.lower() == department.lower()]
    return {
        "employees": [e.to_dict() for e in employees],
        "stats": registry.stats(),
        "departments": sorted({e.department for e in registry.all_employees()}),
    }


@router.get("/capabilities")
async def list_capabilities(current_user: CurrentUser) -> dict:
    """The taxonomy, plus who covers each capability."""
    index = registry.capability_index()
    return {
        "capabilities": [
            {
                "slug": c.slug, "label": c.label, "domain": c.domain,
                "description": c.description,
                "coveredBy": [e.id for e in index.get(c.slug, [])],
            }
            for c in caps.ALL
        ],
        "domains": sorted({c.domain for c in caps.ALL}),
    }


@router.get("/tools")
async def list_tools(current_user: CurrentUser, db: DB,
                     project_id: uuid.UUID | None = Query(default=None)) -> dict:
    """The Tool Layer, and which connected apps this project can actually reach."""
    connected: dict[str, bool] = {}
    if project_id is not None:
        connected = await toolbelt.available_apps(project_id, current_user.org_id, db)
    return {"tools": [t.to_dict() for t in toolbelt.all_tools()],
            "apps": toolbelt.apps(), "connected": connected}


@router.get("/health")
async def employee_health(current_user: CurrentUser, db: DB,
                          project_id: uuid.UUID = Query(...)) -> dict:
    """Status board: can each employee actually work right now?"""
    return {"employees": await orchestrator.health_report(project_id, current_user.org_id, db)}


@router.get("/{employee_id}")
async def get_employee(employee_id: str, current_user: CurrentUser,
                       version: str | None = Query(default=None)) -> dict:
    employee = registry.get(employee_id, version=version)
    if employee is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Employee not found")
    data = employee.to_dict()
    data["versions"] = registry.versions(employee_id)
    return data


# --- delegation ---------------------------------------------------------------


@router.post("/delegate")
async def delegate(body: DelegateRequest, current_user: CurrentUser, db: DB) -> dict:
    """Hand a goal to the company. The Orchestrator does the rest."""
    if not body.goal.strip():
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Goal is required")
    report = await orchestrator.run(
        body.goal.strip(), body.project_id, current_user.org_id, db,
        persona=body.persona, tier=body.tier, capabilities=body.capabilities,
        parallelism=max(1, min(body.parallelism, 6)),
    )
    return report.to_dict()


@router.post("/plan")
async def preview_plan(body: DelegateRequest, current_user: CurrentUser, db: DB) -> dict:
    """Show the team and the plan without executing -- the approval surface."""
    from app.employees import brand_dna
    from app.employees.context import WorkContext
    from app.services.llm_service import get_org_llm_keys

    keys = await get_org_llm_keys(current_user.org_id, db)
    dna = await brand_dna.build(body.project_id, current_user.org_id, db)
    ctx = WorkContext(goal=body.goal, project_id=body.project_id, org_id=current_user.org_id,
                      db=db, dna=dna, persona=body.persona, keys=keys)
    wanted = body.capabilities or await orchestrator.understand(body.goal, body.persona, ctx)
    tasks = orchestrator.build_plan(wanted, body.goal, ctx)
    team = orchestrator.team_for(tasks)
    return {
        "capabilities": wanted,
        "tasks": [t.to_dict() for t in tasks],
        "layers": [[t.id for t in layer] for layer in orchestrator.layers(tasks)],
        "team": [{"id": e.id, "name": e.name, "role": e.role, "department": e.department,
                  "icon": e.icon} for e in team],
        "brandDna": dna.to_dict(),
        "logs": ctx.execution_log(),
    }


# --- institutional memory -----------------------------------------------------


@router.get("/memory/recall")
async def recall_memory(current_user: CurrentUser, db: DB,
                        project_id: uuid.UUID = Query(...),
                        q: str = Query(default=""),
                        employee_id: str | None = Query(default=None),
                        scope: str = Query(default="project"),
                        limit: int = Query(default=20, le=100)) -> dict:
    hits = await memory_layer.recall(db, org_id=current_user.org_id, project_id=project_id,
                                     query=q, employee_id=employee_id, scope=scope, limit=limit)
    return {"memories": [h.to_dict() for h in hits]}


@router.post("/memory")
async def write_memory(body: MemoryWrite, current_user: CurrentUser, db: DB) -> dict:
    employee = registry.get(body.employee_id)
    memory_id = await memory_layer.remember(
        db, org_id=current_user.org_id, project_id=body.project_id,
        employee_id=body.employee_id, content=body.content, scope=body.scope,
        kind=body.kind, key=body.key,
        department=employee.department if employee else None)
    if memory_id is None:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,
                            detail="Memory could not be stored")
    return {"id": memory_id}


@router.delete("/memory/{memory_id}")
async def delete_memory(memory_id: uuid.UUID, current_user: CurrentUser, db: DB) -> dict:
    ok = await memory_layer.forget(db, org_id=current_user.org_id, memory_id=memory_id)
    if not ok:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Memory not found")
    return {"ok": True}

File: apps/web/lib/employees.ts
/** The AI company, as the frontend sees it.
 *
 *  Nothing here hardcodes the roster. The registry on the API is the single
 *  source of truth; this module types it, fetches it, and maps the icon name
 *  each employee declares onto a real component. An employee hired on the
 *  backend appears in the UI with no frontend change.
 */
import {
  BarChart3, Compass, Footprints, Palmtree, Radar, ScrollText, Sparkles,
  Wand2, Wind, type LucideIcon,
} from "lucide-react";
import { apiClient } from "./api";

export interface EmployeeAction {
  id: string;
  label: string;
  description: string;
  capabilities: string[];
  weight: "light" | "heavy";
  inputs: string[];
  outputs: string[];
  requiresApproval: boolean;
  requiresPermissions: string[];
}

export interface Employee {
  id: string;
  name: string;
  codename: string;
  avatar: string;
  icon: string;
  role: string;
  department: string;
  description: string;
  version: string;
  status: "active" | "beta" | "deprecated" | "disabled";
  personality: string;
  expertise: string[];
  goals: string[];
  capabilities: string[];
  allowedTools: string[];
  connectedApps: string[];
  permissions: string[];
  memoryScope: "self" | "department" | "project" | "org";
  knowledgeSources: string[];
  supportedInputs: string[];
  supportedOutputs: string[];
  consumes: string[];
  producesFor: string[];
  actions: EmployeeAction[];
  versions?: string[];
}

export interface RegistryStats {
  employees: number;
  active: number;
  departments: number;
  actions: number;
  capabilities: number;
}

export interface EmployeeRegistry {
  employees: Employee[];
  stats: RegistryStats;
  departments: string[];
}

export interface CapabilityInfo {
  slug: string;
  label: string;
  domain: string;
  description: string;
  coveredBy: string[];
}

export interface EmployeeTool {
  name: string;
  label: string;
  description: string;
  kind: "data" | "app";
  app: string;
  permission: string;
  writes: boolean;
}

export interface EmployeeHealth {
  id: string;
  name: string;
  department: string;
  status: string;
  ok: boolean;
  detail: string;
  checks: Record<string, unknown>;
  connectedApps: Record<string, boolean>;
}

export interface PlannedTask {
  id: string;
  goal: string;
  capabilities: string[];
  employeeId: string | null;
  actionId: string | null;
  dependsOn: string[];
  why: string;
}

export interface PlanPreview {
  capabilities: string[];
  tasks: PlannedTask[];
  /** Task ids grouped by execution layer — each layer runs in parallel. */
  layers: string[][];
  team: { id: string; name: string; role: string; department: string; icon: string }[];
  brandDna: Record<string, unknown>;
  logs: { at: string; level: string; event: string; detail: Record<string, unknown> }[];
}

export interface RunReport {
  goal: string;
  ok: boolean;
  team: string[];
  tasks: (PlannedTask & { ok: boolean; summary: string; error: string | null })[];
  artifacts: Record<string, unknown>[];
  logs: { at: string; level: string; event: string; detail: Record<string, unknown> }[];
  cost: { calls: number; by_model: Record<string, number> };
  error: string | null;
}

export interface MemoryEntry {
  id: string;
  employeeId: string;
  scope: string;
  kind: string;
  key: string | null;
  content: string;
  meta: Record<string, unknown>;
  score: number;
}

/** Icon names the backend may declare, mapped to components.
 *  Unknown names fall back to a neutral mark rather than breaking the page. */
const ICONS: Record<string, LucideIcon> = {
  radar: Radar,
  "scroll-text": ScrollText,
  wind: Wind,
  "wand-2": Wand2,
  footprints: Footprints,
  palmtree: Palmtree,
  compass: Compass,
  "bar-chart": BarChart3,
  sparkles: Sparkles,
};

export function employeeIcon(name: string): LucideIcon {
  return ICONS[name] ?? Sparkles;
}

/** Department accent tokens. Uses CSS variables so palettes stay per-project. */
export const DEPARTMENT_ACCENT: Record<string, string> = {
  Strategy: "text-primary bg-primary/10",
  Content: "text-amber-500 bg-amber-500/10",
  Marketing: "text-sky-500 bg-sky-500/10",
  "Creative Studio": "text-violet-500 bg-violet-500/10",
  Intelligence: "text-rose-500 bg-rose-500/10",
  Research: "text-emerald-500 bg-emerald-500/10",
  Growth: "text-cyan-500 bg-cyan-500/10",
};

export function departmentAccent(department: string): string {
  return DEPARTMENT_ACCENT[department] ?? "text-muted-foreground bg-muted";
}

// --- API ---------------------------------------------------------------------

export async function listEmployees(params?: {
  department?: string;
  capability?: string;
}): Promise<EmployeeRegistry> {
  const q = new URLSearchParams();
  if (params?.department) q.set("department", params.department);
  if (params?.capability) q.set("capability", params.capability);
  const suffix = q.toString() ? `?${q}` : "";
  return apiClient.get<EmployeeRegistry>(`/employees${suffix}`);
}

export async function getEmployee(id: string): Promise<Employee> {
  return apiClient.get<Employee>(`/employees/${id}`);
}

export async function listCapabilities(): Promise<{
  capabilities: CapabilityInfo[];
  domains: string[];
}> {
  return apiClient.get(`/employees/capabilities`);
}

export async function listEmployeeTools(projectId?: string): Promise<{
  tools: EmployeeTool[];
  apps: string[];
  connected: Record<string, boolean>;
}> {
  const suffix = projectId ? `?project_id=${projectId}` : "";
  return apiClient.get(`/employees/tools${suffix}`);
}

export async function employeeHealth(projectId: string): Promise<{ employees: EmployeeHealth[] }> {
  return apiClient.get(`/employees/health?project_id=${projectId}`);
}

/** Preview the team and plan without spending anything — the approval surface. */
export async function previewPlan(body: {
  goal: string;
  project_id: string;
  persona?: string;
  capabilities?: string[];
}): Promise<PlanPreview> {
  return apiClient.post<PlanPreview>("/employees/plan", body);
}

/** Hand a goal to the company and let the Orchestrator deliver it. */
export async function delegate(body: {
  goal: string;
  project_id: string;
  persona?: string;
  tier?: string;
  capabilities?: string[];
  parallelism?: number;
}): Promise<RunReport> {
  return apiClient.post<RunReport>("/employees/delegate", body);
}

export async function recallMemory(params: {
  project_id: string;
  q?: string;
  employee_id?: string;
  limit?: number;
}): Promise<{ memories: MemoryEntry[] }> {
  const q = new URLSearchParams({ project_id: params.project_id });
  if (params.q) q.set("q", params.q);
  if (params.employee_id) q.set("employee_id", params.employee_id);
  if (params.limit) q.set("limit", String(params.limit));
  return apiClient.get(`/employees/memory/recall?${q}`);
}

export async function forgetMemory(memoryId: string): Promise<{ ok: boolean }> {
  return apiClient.delete(`/employees/memory/${memoryId}`);
}
